### DISCUSSION TITLE
Async executor

### DISCUSSION THREAD

**Discussion by a-meledin** [apache/airflow#35642] [General]
### Description

Problem: 
- When a task logic is written in another language (C/C#/etc) and Airflow is used just for: (1) traversing a Dag; (2) executing remote logic over SSH/REST API; (3) taking logs or their summary back through SFTP/other methods ... In such a setting Local/CeleryExcutor multiprocessing logic gives too much overhead. 
- E.g. we have a production DAG with 1500 task instances to run in parallel. It requires setting approximately 50 celeryd's with 32 threads each. Resource/CPU consumption as compared to real work done by Airflow in such a case cannot be justified. 

Possible solution: 
- Async multithreaded executor seems to be an option. 
- As each task is executed the same way:
-- @task code is empty or contains only configuration
-- configuration is handled by user defined class that is passed to executor.  

NB: Triggerer is not considered as an option, because the process for ti is spawn several times. This multiplies overhead.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Airflow tasks are by definition and design isolated - each task executes in potentially different machine and different process (each task by definition is run in a forked process). Again this is by design and idea is that the overhead it introduces is negligble comparing to actual task being executed. This is also because every task requires at least few reads and writes to the database. so the overhead of starting a task will always be quite significant.

You seem to want to do somethign similar to Kafka where you have a lot of really small tasks that are executed as fast as possible. This is a non-goal for Airflow and if you are trying to use airflow for that, you probably picked a wrong too. 

However maybe you can optimize some things. - If you would like to take a stab on it, you are free to optmize some aspects of it. I am not sure exactly what yo mean by "async multithreaded executor" when all the tasks are executed via forked processes and database operations are involved, but if you have some ideas there how to change architecture of Airflow and what idea you have in mind, the right forum for that i s not a feature request but an Airflow Improvement Proposal https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals - which should be acompanied by proposal and discussion at the devlist where you should propose, explain, presnet your idea, including consequences and details what you are proposing - feel free to do that if you think you have a sound proposal.

Converted that into discussion in case more discussion is needed.

**Reply by a-meledin** [reply]
The main problem is that we have quite big DAG of 6000 task instances with **non-python code**. Each dagRun in production takes up to 15-24 hours to execute and we have strict SLA on the time with the goal of minimizing execution time. Currently, running an empty 6000 ti DAG takes approximately 1.5-2 hour overhead on resources listed in my initial post.

Because code is **non-python code**, Airflow is used for I/O operations only (exec over SSH/read log file over SFTP). As known, I/O in python is better done in Async manner and not recommended for multiprocessing architecture due to [GIL](http://www.dabeaz.com/python/GIL.pdf). Also, because each Airflow Worker uses the same python code for SSH/SFTP operations, it's better to share it between async threads. That's why "async executor" idea is on the table (Dagster has similar request [#4041](https://github.com/dagster-io/dagster/issues/4041)).

In such "async executor" it seems there are no standalone workers. They are built in into executor. That is also not very optimal, because many SSH connections (in the DAG listed we have up to 1500 parallel task instances) on the same time isn't brilliant choice. This can be optimized if "async executor" once fired a process over SSH in NOHUP mode, falls into Deffered (like in Triggerer), periodically checking for new log entries to read, reading them and going to Deffered again.  A pool of SSH connections can be hold limited in such schema and not overgrown (the same idea is used in pgBouncer).

Is it clear now why this request appeared? I think a good guy in this [discussion](https://github.com/apache/airflow/discussions/24606) was very close to close the need listed above with his idea of async executor based on Trigger functionality. However, he was left unsupported with his idea as I've seen so far.

Best regards. Alexey

**Reply by potiuk** [reply]
Why do you think current overhead is too big?  I do not think you can do a lot better IMHO. And in order to implement something that you are proposing, I think you will have to perform the overhead measurement and write a prototype showing that you will - actually - gain anything. I believe you are under-estimating the power of POSIX fork model we and Celery uses here

I am not sure if you realize but default way of running workers by Celery is to fork the workers. Forking in POSIX system works in the way that it re-uses the memory that has not been modified (copy on write). Which basically means that if you fork Python interpreter late enough to have the Python interpreter import most of the modules it needs to import, there is virtually no overhead (except local stack) in resources used by  even many worker processes that were forked by the master process - because all of the processes basically use the same "read" version of the memory they had when fork started.

And each worker is isolated from each other of course, they have their own interpreter running, their own GIL, their own sockets and files opened (so each writes to its own file)  etc. etc. This architecture optimizes both memory used and allows to have the right isolation and effective multi-processing used because each worker is effectively a seperate process (even if they share vast amount of memory between them). 

And I believe the only overhead that counts in this case is memory. There are virtually no other resources that would be - potentially - problematic with this architecture.

Now - of course - there might be cases and libraries that are causing problems. If your python code woudl start reloading the imported libraries or heavily modify heap memory (and causing a lot of copy-on-write) then the memory overhead would be much bigger - but I believe this is not happening too much in regular Airflow "local task" execution - what we are mostly doing is communicating with the DB and writing logs - which on it's own consittute overhead on sockets and all the stack memory used for it, but in general this is also the same kind of overhead you expect for having such mutliple sockets and connection opened by whatever single-process/async threads when the communication is multiplexed using multiple sockets etc.

So `theorethically' IMHO there are not a lot of gains / overhead that you can get this way.

But if you woudl like to dig deeper @a-meledin , I think the right way for you to continue on that is to gather some data - how much overhead the current architecture introduces, how big overhead it is, and possibly you should  write some proof-of-concept solution showing simiilar behaviour and demonstrating the gains. 

Other than that - this is purely theorethical discussion, where we have unnknown gains of the new hypothetical architecture you want to propose.

**Reply by potiuk** [reply]
> Is it clear now why this request appeared? I think a good guy in this https://github.com/apache/airflow/discussions/24606 was very close to close the need listed above with his idea of async executor based on Trigger functionality. However, he was left unsupported with his idea as I've seen so far.

The guy was not left unsupported. He got the same explanation as I did above and the same request to demonstrate the gains of the proposal. We are not going to heavliy modify assumptions of Airlfow without really knowing if this improves anything (the overhead) - especially that current architecture and the use of forks and POSIX forks is quite a lot focused on optimizing the overhead and we believe that it is quite optimal (even for your case).

But if you (or the other guy) prove otherwise and show actual gains you can get by implementing a proof-of-concept where you can demonstrate the gains, i am all ears.
